### ISSUE TITLE
Extending __all__ error

### ISSUE THREAD

**Issue by jtpavlock** [microsoft/pyright#3835]
**Describe the bug**
Trying to extend `__all__` results in an error.

> warning: Operation on "__all__" is not supported, so exported symbol list may be incorrect (reportUnsupportedDunderAll)

in a parent module's `__init__.py`:
```python
import submodule
from submodule import *

__all__ = []
__all__.extend(submodule.__all__)  # error
```

`submodule.py`
```python
__all__ = ["myClass"]
```

On pyright commandline 1.1.266

**Comment by erictraut** [comment]
I'm not able to repro the problem as reported.

I presume that `submodule.py` also declares (or imports) the type `myClass`, something like this...

```python
__all__ = ["myClass"]

class myClass:
    ...
```

Can you confirm that you are still seeing an error with the above code? If not, is there something else I need to do to repro the problem?

**Comment by jtpavlock** [comment]
Here's a more complete repro:

```
$ tree my_pkg/
my_pkg/
├── __init__.py
└── submodule
    ├── __init__.py
    └── my_var.py
```

`my_pkg/__init__.py`
```python
from submodule import my_var
from submodule.my_var import *

__all__ = []
__all__.extend(my_var.__all__)
```

`submodule/__init__.py`
```
(empty)
```

`my_var.py`
```python
__all__ = ["var"]

var = 1
```

**Comment by erictraut** [comment]
Thanks for the additional repro steps. I was able to reproduce the problem, and I understand the cause. You're doing something here that I haven't seen before. Typically, a submodule would expose an `__all__` within its `__init__.py` source file. In your example, you're attempting to access `__all__` from a sub-submodule. 

The `__all__` mechanism in Python presents significant challenges for all static analysis tools because this symbol can be manipulated in a variety of ways. Static analysis tools don't actually execute the code, so they need to include special-case logic to deal with each case. The logic in this particular case wasn't handling this condition correctly. I've updated the code to better handle the case where you're importing from a sub-submodule. The fix will be included in the next release of pyright.

In the meantime, you can work around the issue by importing `my_var` like this:
```python
import submodule.my_var as my_var
```

**Comment by jtpavlock** [comment]
Thanks for the detailed explanation and fix!

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.267, which I just published. It will also be included in a future version of pylance.
